File: 0_100/24_swap_nodes_in_pairs.py
def swapPairs(head):
    if head is None or head.next is None:
        return head
    
    middle = head.next
    prev = None
    
    while head and head.next: 
        if prev is not None:
            prev.next = head.next
        prev = head
        
        nextNode = head.next.next
        head.next.next = head

        head.next = nextNode
        head = nextNode

    return middle

def swap (head):
    if head is None or head.next is None:
		return head

	second = head.next
	third = second.next

	second.next = head
	head.next = swap(third)

	return second

def swapPairsRecursive(head):
    return swap(head)

File: 0_99/24_swap_nodes_in_pairs.go
package main

func swapPairs(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}

	middle := head.Next
	prev := &ListNode{0, nil}

	for head != nil && head.Next != nil {
		if prev != nil {
			prev.Next = head.Next
		}
		prev = head
		nextNode := head.Next.Next
		head.Next.Next = head

		head.Next = nextNode
		head = nextNode
	}

	return middle
}

func swapPairsRecursive(head *ListNode) *ListNode {
	newHead := swap(head)
	return newHead
}

func swap(node *ListNode) *ListNode {
	if node == nil || node.Next == nil {
		return node
	}

	second := node.Next
	third := second.Next

	second.Next = node
	node.Next = swap(third)

	return second
}
